Add security tutorial (#70)
Knotx/knotx-website: src/render/tutorials/api-gateway-security/2_0.html.md

        .get("/api/secure/basic")
      .then()
        .assertThat()
        .statusCode(401);
  }

  @Test
  @DisplayName("GIVEN authorization WHEN call basicAuth API EXPECT Ok")
  void givenAuthorizationWhenCallBasicAuthApiExpectOk() {
    given()
        .port(8092)
        .header("Authorization", "Basic am9objpzM2NyM3Q=")
      .when()
        .get("/api/secure/basic")
      .then()
        .assertThat()
        .statusCode(200);
  }
}
```

We've created two simple tests using [REST-assured](http://rest-assured.io/). They do the same actions we did above:

- call secured endpoint without authorization and expect `401 - Unauthorized` response
- call the same endpoint with authorization and expect `200 - OK` response

If you try rebuilding the project now (remember to stop any running containers before that) you'll notice that those test will be invoked. Now we're confident that if our future changes will somehow disable or modify our `/api/secure/basic` logic, we'll know at build time!

## JWT

Now let's do something fancier than just passing the credentials. We'll use [JSON Web Tokens](https://jwt.io/).

A JWT (JSON Web Token) is a well-known solution for API authorization.

> The JWT allows verifying the owner of some JSON data. The JSON data stored in a JWT is serialized, not encrypted. So it can be seen by anyone so always use HTTPS.

Reviewer: ```suggestion
The JWT allows verifying the owner of some JSON data. The JSON data stored in a JWT is serialized, not encrypted. So it can be seen by anyone (remember to always use HTTPS).
```

```hocon
routingOperations = ${routingOperations} [
  // ...
  {
    operationId = hello-world-operation-jwt
    handlers = [
      {
        name = hellohandler
        config = {
          message = "Hello World From Knot.x with JWT!"
        }
      }
    ]
  }
]

securityHandlers = [
  // ...
  {
    schema = helloJwtAuth
    factory = helloJwtAuthFactory
    config = {
      algorithm = "HS256"
      publicKey = "M0NTY3ODkwIiwibmFtZSI6"
      symmetric = true
    }
  }
]
```

Again, we're reusing `hellohandler` with some different message without any changes in the code. Authentication / authorization logic is not a part of the routing handler implementation.

As for security handlers, we'll be using a new factory with a configuration as above: `HS256` mentioned above with a `symmetric` key. A signature is generated by calculating a digest using the HMAC-SHA256 hashing algorithm. The key used in HMAC is, by definition, symmetric. A symmetric key is used for both to encrypt and decrypt JWT. So we need to configure two properties: `symmetric` and `publicKey`. More info available [here](https://vertx.io/docs/vertx-auth-jwt/java/#_loading_keys).

> > Take care not to expose your private keys anywhere publicly!

Reviewer: ```suggestion
> Be careful not to disclose your private keys anywhere publicly!
```

4. We map the resulting observable with the token to `fetchUserInfo` which creates a `GET` request to Google API with the bearer token authentication and sends it.
5. Finally we subscribe for the whole processing chain and end our response with a personalized *hello* message!

As always, let's not forget to register our factories in the `resources/META_INF/services` directory!

`io.knotx.server.api.handler.RoutingHandlerFactory`
```
io.knotx.examples.security.routing.GoogleLoginRedirectHandlerFactory
io.knotx.examples.security.routing.OAuth2HelloHandlerFactory
```

### Build & Run

Now let's see how it works.

First, let's build the project (as mentioned before, remember to stop all the running containers!):

```
$ gradlew clean build-docker
```

and run it:

```
$ docker run -p8092:8092 knotx-example/secure-api-gateway
```

Now let's type in `http://localhost:8092/api/secure/login` in your browser.

![Select Google Account](/img/blog/api-gateway-security/security-google-select-account.png)

Select your account.

![Select Google Account](/img/blog/api-gateway-security/security-google-hello-message.png)

And there is our personalized hello message!

## Summary

> In this tutorial we have successfully secured our API Gateway with various authentication methods. We even managed to access secured data from external API using OAuth2 flow.

Reviewer: I think that summary deserves few more sentences after so many happened above.
Maybe write about:
- Security Handlers as way of enriching API with security in Knot.x (using SPI)
- Connecting external security provider with API and Server routing.
- Mention about importance of tests.

Author: Splitted into separate tutorials and added adequate summary there

---
title: API gateway Security
author: Mateusz512
keywords: tutorial
date: 2019-10-28
layout: tutorial
knotxVersions:
  - 2.0.0
---
## Overview

In this tutorial, we will take a look on how to secure an API gateway.

> This tutorial is based on the code we implemented in [Getting Started with Docker tutorial](/tutorials/getting-started-with-docker/). If you didn't complete the tutorial, we highly encourage you to do so.

Author: Actually Configurable API gateway is not needed to understand this tutorial. We're not using any fragments or actions here.

    JsonObject jsonObject = new JsonObject()
        .put("message", message);
    return event -> event.response().end(jsonObject.toString());
  }

// ...
```

In this tutorial we'll make sure that this handler is accessible only for authenticated users.

## Basic auth

Let's start with the most basic way of authenticating our requests - [Basic auth](https://en.wikipedia.org/wiki/Basic_access_authentication).

### Design & Setup

As always, we'll use the [Design First approach](https://swagger.io/blog/api-design/design-first-or-code-first-api-development/), therefore we'll modify our [Open API 3](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md) configuration file. Let's add two new endpoints in the `paths:` section and a new `securityScheme`  in the [`components`](https://swagger.io/docs/specification/components/) section:
 
`knotx/conf/openapi.yaml`
```yaml
// ...

paths:  
  /api/public:  // a public endpoint with no security, renamed from /api/hello
    get:
      operationId: hello-world-operation  // the same operation as in /api/hello
      responses:
        default:
          description: Public Hello World API

>   /api/secure/basic:  // endpoint with basic auth authentication

Author: done

```

The `build.gradle.kts` takes care of any dependencies we'll need in this tutorial: 

```kotlin
plugins {
    `java-library`
}

dependencies {
    "io.knotx:knotx".let { v ->
        implementation(platform("$v-dependencies:${project.property("knotx.version")}"))
        implementation("$v-fragments-handler-api:${project.property("knotx.version")}")
    }
    "io.vertx:vertx".let { v ->
        implementation("$v-core")
        implementation("$v-rx-java2")
        implementation("$v-health-check")
        implementation("$v-auth-shiro")
        implementation("$v-auth-jwt")
        implementation("$v-auth-oauth2")
    }
    "org.apache".let { v ->
        compile("$v.httpcomponents:httpclient:4.5.3")
        compile("$v.commons:commons-lang3:3.9")
    }
}
```

Let's not forget to add the module to the main `settings.gradle.kts`!

```
include("security-module")
// ...
project(":security-module").projectDir = file("modules/security-module")
```

Now let's focus on the `resources`.

Let's get back to the `securityHandlers` we defined before. The configuration we pass to `helloBasicAuthFactory` is a path to user/roles configuration file. 

> In this example we'll be using Vert.x's implementation of [Apache Shiro Auth](https://shiro.apache.org/index.html). Therefore, the configuration must be in a format described [here](https://vertx.io/docs/vertx-auth-shiro/java/#_the_shiro_properties_auth_provider).

Author: done

As mentioned before, we'll create the file in module's `resources` folder and name it `users.properties`.

> Let's create a sample user `john` with a super secure password. He'll be an administrator with all the `role`s:

Author: done

`users.properties`
```
user.john = s3cr3t,administrator
role.administrator=*
```

Now let's take care of the actual implementation:

`BasicAuthHandlerFactory.java`
```java
package io.knotx.examples.security.auth;

import io.knotx.server.api.security.AuthHandlerFactory;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.auth.shiro.ShiroAuthOptions;
import io.vertx.reactivex.core.Vertx;
import io.vertx.reactivex.ext.auth.shiro.ShiroAuth;
import io.vertx.reactivex.ext.web.handler.AuthHandler;
import io.vertx.reactivex.ext.web.handler.BasicAuthHandler;

public class BasicAuthHandlerFactory implements AuthHandlerFactory {

  @Override
  public String getName() {
    return "helloBasicAuthFactory";
  }

  @Override
  public AuthHandler create(Vertx vertx, JsonObject config) {
    final ShiroAuth shiroAuth = ShiroAuth.create(vertx, new ShiroAuthOptions().setConfig(config));
    return BasicAuthHandler.create(shiroAuth);
  }
}
```

> It's as simple as creating a [`BasicAuthHandler`](https://vertx.io/docs/apidocs/io/vertx/ext/web/handler/BasicAuthHandler.html) that will use [`ShiroAuth`](https://vertx.io/docs/apidocs/io/vertx/ext/auth/shiro/ShiroAuth.html) instance created with our config.

Author: `ShiroAuth` is a class name

4. Now execute the request.

![Basic Auth Postman response](/img/blog/api-gateway-security/security-postman-basicauth-response.png)

Now the request is successfully authorized and we receive the proper response!

If you don't want to use Postman, you can execute this `curl` command:

```
$ curl -X GET http://localhost:8092/api/secure/basic -H 'Authorization: Basic am9objpzM2NyM3Q='
{"message":"Hello World From Knot.x with Basic Auth!"}
```

### Tests

It's always a good practice to test our code. Let's implement some tests that will automatically do the same things we did manually above.

The tests we're about to write will be functional tests, that will be executed on a real, running Docker image, just after the `healthcheck` (described in [this tutorial](http://knotx.io/tutorials/getting-started-with-docker/2_0/)) executes successfully.

Let's add a new file: `functional/src/test/java/com/project/test/functional/BasicAuthITCase.java`

```java
class BasicAuthITCase {

  @Test
  @DisplayName("GIVEN no authorization WHEN call basicAuth API EXPECT Unauthorized")
  void givenNoAuthorizationWhenCallBasicAuthApiExpectUnauthorized() {

>     // @formatter:off

Author: done

## JWT

Now let's do something fancier than just passing the credentials. We'll use [JSON Web Tokens](https://jwt.io/). 

We'll not get into neither cryptography nor JWT specification details, more information is available [here](https://jwt.io/introduction/). 

We'll use [HMAC-SHA256](https://www.ietf.org/rfc/rfc2104.txt) algorithm with a symmetric key. Ideally, we would want to implement a flow as below:

![JWT flow](/img/blog/api-gateway-security/security-jwt-flow.png)

1. User logs in with their credentials.
2. User receives a token encrypted with a key known only to the Server.
3. User uses the received token to authorize requests to other endpoints. The server uses the key to verify if the token is valid or not.

Since we'll be focusing only on step 3, we'll mimic the result of the other two by generating the token ourselves.

But let's start with design.

### Design & Setup

We'll need a new endpoint and a new security schema:

`knotx/conf/openapi.yaml`
```yaml
// ...

paths:  
  // ...
  /api/secure/jwt:
    get:
      operationId: hello-world-operation-jwt
      security:

>           - helloJwtAuth: []

Author: It's explained in the linked documentation. Should I ~~copypaste~~ quote it here?

      responses:
        default:
          description: Hello World API protected with JWT          
  // ...
  
components:
  securitySchemes:
    // ...
    helloJwtAuth:
      type: http
      scheme: bearer
      bearerFormat: JWT
```

Again, we defined a new endpoint and attached security schema with JWT bearer auth.

Now, let's take care of our new operation `hello-world-operation-jwt` and security scheme `helloJwtAuth`:

`knotx/conf/routing/operations.conf`
```hocon
routingOperations = ${routingOperations} [
  // ...  
  {
    operationId = hello-world-operation-jwt
    handlers = [
      {
        name = hellohandler
        config = {
          message = "Hello World From Knot.x with JWT!"
        }
      }
    ]
  }
]

securityHandlers = [
  // ...
  {
    schema = helloJwtAuth
    factory = helloJwtAuthFactory
    config = {
      algorithm = "HS256"
      publicKey = "M0NTY3ODkwIiwibmFtZSI6"
      symmetric = true
    }
  }
]
```

Again, we're reusing `hellohandler` with some different message. 

> As for security handlers, we'll be using a new factory with a configuration as above: `HS256` mentioned above with a `symmetric` key. Although we specify it as a `publicKey`, it's a private key. This naming confusion is  just a requirement for the Vert.x config we'll use. More info available [here](https://vertx.io/docs/vertx-auth-jwt/java/#_loading_keys). 

Author: Actually that's a really bizarre behaviour of Vert.x's JWT implementation and if the key is configured as symmetric it's using the public one.

> As for security handlers, we'll be using a new factory with a configuration as above: `HS256` mentioned above with a `symmetric` key. Although we specify it as a `publicKey`, it's a private key. This naming confusion is  just a requirement for the Vert.x config we'll use. More info available [here](https://vertx.io/docs/vertx-auth-jwt/java/#_loading_keys). 

Reviewer 2: Maybe we can write something like:

The default signature method for JWT’s is known as HS256. A signature is generated by calculating a digest using the HMAC-SHA256 hashing algorithm. The key used in HMAC is, by definition, symmetric. A symmetric key is used for both to encrypt and decrypt JWT. So we need to configure two properties: `symmetric` and `publicKey`.

Please note that the key is shared between the authentication provider (in this example we would generate the digest by our own) and the 3rd parties (the secured API).

    JWTAuthOptions jwtAuthOptions = new JWTAuthOptions().addPubSecKey(pubSecKey);
    return JWTAuthHandler.create(JWTAuth.create(vertx, jwtAuthOptions));
  }
}
```

We're creating a configuration based on the config JSON the factory receives from `operations.conf` and return a [`JWTAuthHandler`](https://vertx.io/docs/apidocs/io/vertx/ext/web/handler/JWTAuthHandler.html) that does all the authentication for us.

As always, let's not forget to register our handler factory:

`io.knotx.server.api.security.AuthHandlerFactory`
```
io.knotx.examples.security.auth.JwtAuthHandlerFactory
```

### Build & Run

Now let's rebuild the project.

> Remember to stop any running containers! If there is a container that uses the `8092` port we're using, our container will not start. You can use this command to stop all running containers: `docker stop $(docker ps -a -q)`

Let's build the image:

```
$ gradlew clean build-docker
```

and run it:

```
$ docker run -p8092:8092 knotx-example/secure-api-gateway
```

After a while the Docker container should be up and running.

Let's try accessing our endpoint:

```
$ curl -X GET http://localhost:8092/api/secure/jwt
Unauthorized
```

> Oops! We were not authorized to access this endpoint. 

Author: done

We need an `Authorization` header for our request. As mentioned before, we'll mimic a successful login and we'll generate the token ourselves.

There are many ways to generate a token. We'll use [`jwt.io`](https://jwt.io/). 

Open the page, scroll to the Debugger section and generate the token:

![Generating JWT using jwt.io](/img/blog/api-gateway-security/security-generate-jwt.png)

1. Select `HS256` as algorithm.
2. Fill in the key you used in configuration.
3. Copy the generated token.

> Let's put the token in the authorization header with the following `curl` command:

Author: done

    String nonce = RandomStringUtils.random(20);

    if (StringUtils.isAnyBlank(redirectUri, scope, clientId)) {
      throw new IllegalArgumentException("Configuration for Google Auth must include redirectUri, clientId and scope");
    }

    return new URIBuilder(AUTH_URL)
        .addParameter("redirect_uri", redirectUri)
        .addParameter("scope", scope)
        .addParameter("client_id", clientId)
        .addParameter("state", state)
        .addParameter("nonce", nonce)
        .addParameter("prompt", "consent")
        .addParameter("access_type", "offline")
        .addParameter("response_type", "code")
        .build().toString();
  }
}
```

As described earlier, this handler will result in a redirect response. We're building the redirect URL based on [this documentation](https://developers.google.com/identity/protocols/OAuth2WebServer#creatingclient):

- `redirect_uri` is the URI that will serve as a redirect from the consent screen after successful login and consent click
- `scope` is used to define which part of the user's resources we want to access (more info [here](https://oauth.net/2/scope/))
- `client_id` is the client id we received in the previous chapter

> - `state` and `nonce` are some random string values, see [this documentation](https://developers.google.com/identity/protocols/OAuth2WebServer#creatingclient) for further details

Reviewer 3: Why are they random? Is it because they are required but we don't need them?

Author: provided more description

  }

  private Single<HttpResponse<Buffer>> fetchUserInfo(String accessToken) {
    return webClient.getAbs(USERINFO_URL)
        .bearerTokenAuthentication(accessToken)
        .rxSend();
  }

}

```

Let's take a look at this file step by step:

1. The constructor is quite straightforward, we create a new [`webClient`](https://vertx.io/docs/vertx-web-client/java/) and save the config for later use.
2. When a request hits our handler, the `handle` method is invoked. We try to obtain the authorization code from the request.
3. Then we try to `exchangeCodeForToken`. The first step is to `prepareTokenRequest`. This time we have to provide the `code`, `client_id`, `client_secret`, `redirect_uri`, `grant_type` parameters and `Content-Length` header (this will be a `POST` request). For more details on those parameters see [this documentation](https://developers.google.com/identity/protocols/OAuth2WebServer#exchange-authorization-code). We send the request using `rxSend` method (for more info about Vert.x's RxJava 2 API check [this documentation](https://vertx.io/docs/vertx-web-client/java/#_rxjava_2_api)) and map the response to the acquired access token.

> 4. We map the resulting observable with the token to `fetchUserInfo` whick creates a `GET` request to Google API with bearer token authentication and sends it.

Reviewer 3: ```suggestion
4. We map the resulting observable with the token to `fetchUserInfo` which creates a `GET` request to Google API with the bearer token authentication and sends it.
```

> 5. Finally we subscribe for the whole processing chain and end our response with a personalized hello message!

Reviewer 3: Before it was `*hello* message`

Author: done

As always, let's not forget to register our factories in the `resources/META_INF/services` directory!
`io.knotx.server.api.handler.RoutingHandlerFactory`
```
io.knotx.examples.security.routing.GoogleLoginRedirectHandlerFactory
io.knotx.examples.security.routing.OAuth2HelloHandlerFactory
```

### Build & Run

Now let's see how it works.

First, let's build the project (as mentioned before, remember to stop all the running containers!):

```
$ gradlew clean build-docker
```

and run it:

```
$ docker run -p8092:8092 knotx-example/secure-api-gateway
```

Now let's type in `http://localhost:8092/api/secure/login` in your browser.

![Select Google Account](/img/blog/api-gateway-security/security-google-select-account.png)

Select your account.

![Select Google Account](/img/blog/api-gateway-security/security-google-hello-message.png)

And there is our personalized hello message!

---

> You can find full project implementation [here](https://github.com/Knotx/knotx-example-project/tree/master/api-gateway/security).

Reviewer 3: How about some summary of what we actually did and what we learned and why !x is awesome etc

Author: I've provided some, let me know if it is satisfactory

---
title: API gateway Security
author: Mateusz512
keywords: tutorial
date: 2019-10-28
layout: tutorial
knotxVersions:
  - 2.0.0
---
## Overview

In this tutorial, we will take a look on how to secure an API gateway.

This tutorial is based on the code we implemented in [Getting Started with Docker tutorial](/tutorials/getting-started-with-docker/). If you didn't complete the tutorial, we highly encourage you to do so.

### Prerequisites

- JDK 8
- Docker (not required, you can use Stack distribution as well)

> - [Postman](https://www.getpostman.com/) (it will simplify things, but `curl` is fine as well) 

Author: I've removed postman

- Modern web browser
- Google account

To make things simple, we'll start from the code already implemented in the tutorial mentioned before, which is available [here](https://github.com/Knotx/knotx-example-project/tree/master/distribution/getting-started).

Let's quickly recap on the things we'll reuse here.

We already implemented the `HelloWorldHandlerFactory` which simply prints the configured message to the HTTP response:

```java
// ...

  @Override
  public Handler<RoutingContext> create(Vertx vertx, JsonObject config) {
    String message = config.getString("message", "Some default message");
    JsonObject jsonObject = new JsonObject()
        .put("message", message);
    return event -> event.response().end(jsonObject.toString());
  }

// ...
```

In this tutorial we'll make sure that this handler is accessible only for authenticated users.

## Basic auth

Let's start with the most basic way of authenticating our requests - [Basic auth](https://en.wikipedia.org/wiki/Basic_access_authentication).

> ### Design & Setup

Reviewer 2: ```suggestion
### Configure basic auth in OpenAPI
```

As always, we'll use the [Design First approach](https://swagger.io/blog/api-design/design-first-or-code-first-api-development/), therefore we'll modify our [Open API 3](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.2.md) configuration file. Let's add two new endpoints in the `paths:` section and a new `securityScheme`  in the [`components`](https://swagger.io/docs/specification/components/) section:
 
`knotx/conf/openapi.yaml`
```yaml
// ...

paths:  
  /api/public:  // a public endpoint with no security, renamed from /api/hello
    get:
      operationId: hello-world-operation  // the same operation as in /api/hello
      responses:
        default:
          description: Public Hello World API
  /api/secure/basic:  // endpoint with basic authentication authentication
    get:
      operationId: hello-world-operation-basic
      security:
        - helloBasicAuth: []  // here we specify which security scheme we'll be using (see below)
      responses:

>         default:

Reviewer 2: Can we specify here responses? There should be `200` and `401` responses. See https://swagger.io/docs/specification/authentication/basic-authentication/.

Reviewer 2: It is explained in so you can link this tutorial: https://knotx.io/tutorials/openapi-and-swagger-ui/2_0/

          description: Hello World API protected with Basic Auth
          
  // ...
  
components:
  securitySchemes:
    helloBasicAuth: // our custom security scheme ...
      type: http  // ... with a http type ...

>       scheme: basic   // ... with a Basic Auth scheme

Author: I've left the comments, but added the description

```

For more information on `basic` scheme see [this documentation](https://swagger.io/docs/specification/authentication/basic-authentication/).

Now we'll take care of handling our operations and security schemas. Let's modify `knotx/conf/routes/operations.conf`: 

```hocon
routingOperations = ${routingOperations} [
  {

>     operationId = hello-world-operation

Reviewer 2: This can be removed.

    handlers = [
      {
        name = hellohandler
        config = {
          message = "Hello World From Knot.x!"
        }
      }
    ]
  }
  {

>     operationId = hello-world-operation-basic

Reviewer 2: There are no changes in operations, can we skip it?

    handlers = [
      {
        name = hellohandler
        config = {
          message = "Hello World From Knot.x with Basic Auth!"
        }
      }
    ]
  }
]

> securityHandlers = [

Reviewer 2: This should be moved to the next section (AuthHandler configuration).

  {
    schema = helloBasicAuth
    factory = helloBasicAuthFactory
    config = {
      properties_path = "classpath:basicauth/users.properties"
    }
  }
]
``` 

As we can see, we reuse our `hellohandler` implemented in `HelloWorldHandlerFactory`, but with different messages each time. 

We also provided an array of [`securityHandlers`](https://github.com/Knotx/knotx-server-http#routing-security). It's a collection of objects which map `schema` with `factory` that must implement [`AuthHandlerFactory`](https://github.com/Knotx/knotx-server-http/tree/master/api#creating-auth-handler) interface. We can also pass some config here (more on that later).

It is worth noting that operations are not aware of any security they will be behind of.

> ### New module

Reviewer 2: ```suggestion
### Authentication Handler implementation
```

For the purpose of this tutorial let's add a new module to our project. Let's name it `security-module` and give it the following structure.:

```
modules
└─ security-module
    ├── build.gradle.kts                                                    // gradle build script
    └── src 
        └── main                   
            ├── java/io/knotx/examples/security/auth
            │     └── BasicAuthHandlerFactory.java                          // the handler factory  
            └── resources
                  ├── basicauth
                  │   └── users.properties                                  // our config file 
                  └── META-INF/services
                      └── io.knotx.server.api.security.AuthHandlerFactory   // META-INF file used by Knot.x to find the handler   
```

The `build.gradle.kts` takes care of any dependencies we'll need in this tutorial: 

```kotlin

> plugins {

Reviewer 2: This can be removed.

  }
}
```

It's as simple as creating a [`BasicAuthHandler`](https://vertx.io/docs/apidocs/io/vertx/ext/web/handler/BasicAuthHandler.html) that will use [`ShiroAuth`](https://vertx.io/docs/apidocs/io/vertx/ext/auth/shiro/ShiroAuth.html) instance created with our config.

Let's not forget to register our factory for [Java ServiceLoader](https://docs.oracle.com/javase/7/docs/api/java/util/ServiceLoader.html):

`io.knotx.server.api.security.AuthHandlerFactory`
```
io.knotx.examples.security.auth.BasicAuthHandlerFactory
```

### Build & Run

In this tutorial we'll be using the Docker distribution, but it will work with Stack distribution as well. Please refer to [this tutorial](http://knotx.io/tutorials/getting-started-with-knotx-stack/2_0/) on how to work with the Stack distribution.

First, let's rename the Docker image we're about to create:

`gradle.properties`
```
// ...
docker.image.name=knotx-example/secure-api-gateway
// ...
```

Now let's build the image:

```
$ gradlew clean build-docker
```

and run it:

```
$ docker run -p8092:8092 knotx-example/secure-api-gateway
```

After a while the Docker container should be up and running.

> Let's try accessing our endpoints:

Reviewer 2: We can skip this.

```
$ curl -X GET http://localhost:8092/api/public
{"message":"Hello World From Knot.x!"}
```

Our public endpoint is working as expected.

Now let's try the other one:

```
$ curl -X GET http://localhost:8092/api/secure/basic
Unauthorized
```

Oops! We were not authorized to access this endpoint. 

Let's open Postman and construct a request there.

![Basic Auth Postman setup](/img/blog/api-gateway-security/security-postman-basicauth.png)

1. Let's create a new `GET` request for the same URL as before.
2. Open the `Authorization` tab.
3. Select `Basic Auth` as the authorization type and type in `john`'s credentials.
4. Now execute the request.

![Basic Auth Postman response](/img/blog/api-gateway-security/security-postman-basicauth-response.png)

Now the request is successfully authorized and we receive the proper response!

> If you don't want to use Postman, you can execute this `curl` command:

Reviewer 2: I would explain this: https://docs.apigee.com/api-platform/system-administration/basic-auth#basic-authentication-format

> You can pass your credentials as a Base64-encoded header or as parameters in an HTTP client.
> When you pass your credentials in the header, you must Base64-encode them. The following is an example of an encoded HTTP Basic Authentication header...

Then you would be able to remove postman at all. I think `curl` examples are enough.

```
$ curl -X GET http://localhost:8092/api/secure/basic -H 'Authorization: Basic am9objpzM2NyM3Q='
{"message":"Hello World From Knot.x with Basic Auth!"}
```

> ### Tests

Reviewer 2: ```suggestion
### Functional tests
```

    given()
        .port(8092)
      .when()
        .get("/api/secure/basic")
      .then()
        .assertThat()
        .statusCode(401);
  }

  @Test
  @DisplayName("GIVEN authorization WHEN call basicAuth API EXPECT Ok")
  void givenAuthorizationWhenCallBasicAuthApiExpectOk() {
    given()
        .port(8092)
        .header("Authorization", "Basic am9objpzM2NyM3Q=")
      .when()
        .get("/api/secure/basic")
      .then()
        .assertThat()
        .statusCode(200);
  }
}
```

We've crated two simple tests using [REST-assured](http://rest-assured.io/). They do the same actions we did above:

- call secured endpoint without authorization and expect `401 - Unauthorized` response
- call the same endpoint with authorization and expect `200 - OK` response

If you try rebuilding the project now (remember to stop any running containers before that) you'll notice that those test will be invoked. Now we're confident that if our future changes will somehow disable or modify our `/api/secure/basic` logic, we'll know at build time!

## JWT

Now let's do something fancier than just passing the credentials. We'll use [JSON Web Tokens](https://jwt.io/). 

> We'll not get into neither cryptography nor JWT specification details, more information is available [here](https://jwt.io/introduction/). 

Reviewer 2: ```suggestion
In this tutorial, we'll not delve into such topics as cryptography or how to store JWT inside the user's browser. We put JWT in the HTTP request headers. In the production, JWT should be stored in an httpOnly cookie, and then can be rewritten by some Apache server to headers.
```

> We'll use [HMAC-SHA256](https://www.ietf.org/rfc/rfc2104.txt) algorithm with a symmetric key. Ideally, we would want to implement a flow as below:

Reviewer 2: This information should be later.
```suggestion 
Ideally, we would want to implement a flow as below:
```

![JWT flow](/img/blog/api-gateway-security/security-jwt-flow.png)

> 1. User logs in with their credentials.

Reviewer 2: ```suggestion
1. A user logs in with his/her credentials.
```

> 2. User receives a token encrypted with a key known only to the Server.

Reviewer 2: ```suggestion
2. The user receives an encrypted JWT in a cookie tagged with the HttpOnly flag (it tells that only the server can access this particular cookie).
```

3. User uses the received token to authorize requests to other endpoints. The server uses the key to verify if the token is valid or not.

Since we'll be focusing only on step 3, we'll mimic the result of the other two by generating the token ourselves.

> But let's start with design.

Reviewer 2: ```suggestion
Let's start with setting our secured API.
```

### Design & Setup

We'll need a new endpoint and a new security schema:

`knotx/conf/openapi.yaml`
```yaml
// ...

paths:  
  // ...
  /api/secure/jwt:
    get:
      operationId: hello-world-operation-jwt
      security:
          - helloJwtAuth: []
      responses:
        default:
          description: Hello World API protected with JWT          
  // ...
  
components:
  securitySchemes:
    // ...
    helloJwtAuth:
      type: http
      scheme: bearer
      bearerFormat: JWT
```

Again, we defined a new endpoint and attached security schema with JWT bearer auth.

Now, let's take care of our new operation `hello-world-operation-jwt` and security scheme `helloJwtAuth`:

`knotx/conf/routing/operations.conf`
```hocon
routingOperations = ${routingOperations} [
  // ...  
  {
    operationId = hello-world-operation-jwt
    handlers = [
      {

>         name = hellohandler

Reviewer 2: Please note that we reuse the same hello handler, without any changes in the code. Authentication / authorization logic is not a part of the routing handler implementation.

        config = {
          message = "Hello World From Knot.x with JWT!"
        }
      }
    ]
  }
]

> securityHandlers = [

Reviewer 2: Move to the end of auth handler implementation section.

> Take care not to expose your private keys anywhere publicly!

### Implementation

Now let's implement our handler factory. Create a class in the `security-module` we created earlier, right next to `BasicAuthHandlerFactory.java`:

`JwtAuthHandlerFactory.java`
```java
package io.knotx.examples.security.auth;

import io.knotx.server.api.security.AuthHandlerFactory;
import io.vertx.core.json.JsonObject;
import io.vertx.ext.auth.PubSecKeyOptions;
import io.vertx.ext.auth.jwt.JWTAuthOptions;
import io.vertx.reactivex.core.Vertx;
import io.vertx.reactivex.ext.auth.jwt.JWTAuth;
import io.vertx.reactivex.ext.web.handler.AuthHandler;
import io.vertx.reactivex.ext.web.handler.JWTAuthHandler;

public class JwtAuthHandlerFactory implements AuthHandlerFactory {

  @Override
  public String getName() {
    return "helloJwtAuthFactory";
  }

  @Override
  public AuthHandler create(Vertx vertx, JsonObject config) {
    PubSecKeyOptions pubSecKey = new PubSecKeyOptions(config);
    JWTAuthOptions jwtAuthOptions = new JWTAuthOptions().addPubSecKey(pubSecKey);
    return JWTAuthHandler.create(JWTAuth.create(vertx, jwtAuthOptions));
  }
}
```

> We're creating a configuration based on the config JSON the factory receives from `operations.conf` and return a [`JWTAuthHandler`](https://vertx.io/docs/apidocs/io/vertx/ext/web/handler/JWTAuthHandler.html) that does all the authentication for us.

Reviewer 2: Remember to add `jwt` in dependnecies.

As always, let's not forget to register our handler factory:

`io.knotx.server.api.security.AuthHandlerFactory`
```
io.knotx.examples.security.auth.JwtAuthHandlerFactory
```

### Build & Run

Now let's rebuild the project.

> Remember to stop any running containers! If there is a container that uses the `8092` port (the one we're using), our container will not start. You can use this command to stop all running containers: `docker stop $(docker ps -a -q)`

Let's build the image:

```
$ gradlew clean build-docker
```

and run it:

```
$ docker run -p8092:8092 knotx-example/secure-api-gateway
```

After a while the Docker container should be up and running.

Let's try accessing our endpoint:

```
$ curl -X GET http://localhost:8092/api/secure/jwt
Unauthorized
```

We were not authorized to access this endpoint. 

We need an `Authorization` header for our request. As mentioned before, we'll mimic a successful login and we'll generate the token ourselves.

There are many ways to generate a token. We'll use [`jwt.io`](https://jwt.io/). 

Open the page, scroll to the Debugger section and generate the token:

> ![Generating JWT using jwt.io](/img/blog/api-gateway-security/security-generate-jwt.png)

Author: I don't think that's necessary
